File: frontend/src/operations/csv-imports.ts
import { mockUsersRepository } from '../administration/users/mock-users-repository'
import { emptyUserDraft, validateUserDraft, type UserDraft } from '../administration/users/users-model'
import { banksProductsRepository } from '../administration/banks-products/banks-products'
import { productStagesRepository } from '../administration/banks-products/product-stages'
import { caseRepository } from '../cases/case-repository'
import { requireAction } from './permissions'

export type ImportKind = 'attendance' | 'case-stage' | 'users'
export interface ImportRow { line: number; values: Record<string, string>; errors: string[] }
export interface ImportPreview { id?: string; kind: ImportKind; rows: ImportRow[]; valid: number; invalid: number; matchedCases: number; unmatchedCases: number }
export interface ImportResult { id: string; kind: ImportKind; at: string; actorId: string; success: number; failed: number; details: string[] }
export interface AttendanceRecord { id: string; userId: string; date: string; status: 'Present' | 'Absent' | 'Late'; importedAt: string; importedBy: string }
export const IMPORT_RESULTS_KEY = 'amafh-v2.mock-import-results.v1'
export const ATTENDANCE_KEY = 'amafh-v2.mock-attendance.v1'
const headers: Record<ImportKind, string[]> = {
  attendance: ['userEmail', 'date', 'status'],
  'case-stage': ['caseNumber', 'stage'],
  users: ['fullName', 'email', 'userType', 'organization', 'organizationScope', 'officeBranch', 'department', 'team'],
}
export function parseCsv(text: string): string[][] {
  const rows: string[][] = []; let row: string[] = []; let cell = ''; let quoted = false
  const source = text.replace(/^\uFEFF/, '')
  for (let index = 0; index < source.length; index++) {
    const char = source[index]
    if (quoted) { if (char === '"' && source[index + 1] === '"') { cell += '"'; index++ } else if (char === '"') quoted = false; else cell += char }
    else if (char === '"' && !cell) quoted = true
    else if (char === ',') { row.push(cell); cell = '' }
    else if (char === '\n' || char === '\r') { if (char === '\r' && source[index + 1] === '\n') index++; row.push(cell); if (row.some(value => value.trim())) rows.push(row); row = []; cell = '' }
    else cell += char
  }
  if (quoted) throw new Error('CSV contains an unclosed quoted value.')
  if (cell || row.length) { row.push(cell); if (row.some(value => value.trim())) rows.push(row) }
  return rows
}
export async function readCsvFile(file: File) {
  if (!file.name.toLowerCase().endsWith('.csv') || file.size > 2 * 1024 * 1024 || file.size === 0) throw new Error('Choose a non-empty CSV file no larger than 2 MB.')
  const text = await file.text()
  if (text.includes('\0')) throw new Error('Binary files are not supported.')
  return text
}
function readArray<T>(key: string): T[] {
  const raw = localStorage.getItem(key)
  if (raw === null) return []
  try { const value: unknown = JSON.parse(raw); if (Array.isArray(value)) return value as T[] } catch { /* handled below */ }
  throw new Error('Local import data could not be read.')
}
function writeArray<T>(key: string, rows: T[]) { try { localStorage.setItem(key, JSON.stringify(rows)) } catch { throw new Error('Could not save local import results.') } }
function validDate(value: string) { return /^\d{4}-\d{2}-\d{2}$/.test(value) && new Date(`${value}T00:00:00Z`).toISOString().slice(0, 10) === value }
export async function previewImport(kind: ImportKind, csv: string): Promise<ImportPreview> {
  const parsed = parseCsv(csv)
  if (!parsed.length || parsed[0].map(value => value.trim()).join('|') !== headers[kind].join('|')) throw new Error(`CSV headers must be: ${headers[kind].join(', ')}`)
  if (parsed.length > 1001) throw new Error('Import at most 1000 rows at a time.')
  const [users, cases, catalogue, existingAttendance] = await Promise.all([mockUsersRepository.list(), caseRepository.list(), banksProductsRepository.list(), Promise.resolve(readArray<AttendanceRecord>(ATTENDANCE_KEY))])
  const stages = (await Promise.all(catalogue.filter(item => item.kind === 'products').map(item => productStagesRepository.list(item.id)))).flat()
  const seen = new Set<string>()
  let matchedCases = 0; let unmatchedCases = 0
  const rows: ImportRow[] = parsed.slice(1).map((cells, index) => {
    const values = Object.fromEntries(headers[kind].map((header, position) => [header, cells[position]?.trim() ?? '']))
    const errors: string[] = []
    if (cells.length !== headers[kind].length) errors.push('Incorrect column count.')
    if (kind === 'attendance') {
      const user = users.find(item => item.email.toLowerCase() === values.userEmail.toLowerCase())
      if (!user) errors.push('User email was not found.')
      if (!/^\d{4}-\d{2}-\d{2}$/.test(values.date) || Number.isNaN(Date.parse(`${values.date}T00:00:00Z`)) || !validDate(values.date)) errors.push('Invalid date.')
      if (!['Present', 'Absent', 'Late'].includes(values.status)) errors.push('Invalid attendance status.')
      const key = `${values.userEmail.toLowerCase()}:${values.date}`
      if (seen.has(key) || existingAttendance.some(item => item.userId === user?.id && item.date === values.date)) errors.push('Duplicate attendance entry.')
      seen.add(key)
    } else if (kind === 'case-stage') {
      const item = cases.find(record => record.caseNumber === values.caseNumber || record.id === values.caseNumber)
      if (!item) { errors.push('Case was not found.'); unmatchedCases++ } else {
        matchedCases++
        const product = catalogue.find(record => record.kind === 'products' && (record.id === item.productId || record.name === item.product && catalogue.some(bank => bank.kind === 'banks' && bank.id === record.parentId && bank.name === item.bank)))
        const stage = stages.find(record => record.productId === product?.id && record.name === values.stage && record.status === 'active')
        if (!stage) errors.push('Stage is not active for this Case Product.')
        if (item.status !== 'SM Approved' || !item.coordinatorId || !item.submittedAt) errors.push('Case is not ready for stage tracking.')
        if (item.stageId === stage?.id) errors.push('Case is already in this Stage.')
      }
      if (seen.has(values.caseNumber)) errors.push('Duplicate Case row.'); seen.add(values.caseNumber)
    } else {
      const email = values.email.toLowerCase()
      const userDraft: UserDraft = { ...emptyUserDraft(), fullName: values.fullName, email, userType: values.userType, organization: values.organization, organizationScope: values.organizationScope as UserDraft['organizationScope'], officeBranch: values.officeBranch, department: values.department, team: values.team }
      errors.push(...Object.values(validateUserDraft(userDraft, users)))
      if (seen.has(email)) errors.push('Duplicate email in CSV.'); seen.add(email)
    }
    return { line: index + 2, values, errors }
  })
  return { kind, rows, valid: rows.filter(row => !row.errors.length).length, invalid: rows.filter(row => row.errors.length).length, matchedCases, unmatchedCases }
}
export const importRepository = {
  async history() { return readArray<ImportResult>(IMPORT_RESULTS_KEY) },
  async attendance() { return readArray<AttendanceRecord>(ATTENDANCE_KEY) },
  async confirm(preview: ImportPreview, actorId: string) {
    await requireAction('Imports', preview.kind, actorId)
    if (!preview.rows.length || preview.valid === 0) throw new Error('No valid rows are available to import.')
    const fresh = await previewImport(preview.kind, [headers[preview.kind].join(','), ...preview.rows.map(row => headers[preview.kind].map(header => `"${row.values[header].replaceAll('"', '""')}"`).join(','))].join('\n'))
    const results: string[] = []; let success = 0
    const [users, cases, catalogue] = await Promise.all([mockUsersRepository.list(), caseRepository.list(), banksProductsRepository.list()])
    const stages = (await Promise.all(catalogue.filter(item => item.kind === 'products').map(item => productStagesRepository.list(item.id)))).flat()
    const attendance = readArray<AttendanceRecord>(ATTENDANCE_KEY)
    for (const row of fresh.rows) {
      if (row.errors.length) { results.push(`Line ${row.line}: ${row.errors.join(' ')}`); continue }
      try {
        if (preview.kind === 'attendance') {
          const user = users.find(item => item.email.toLowerCase() === row.values.userEmail.toLowerCase())!
          attendance.push({ id: crypto.randomUUID(), userId: user.id, date: row.values.date, status: row.values.status as AttendanceRecord['status'], importedAt: new Date().toISOString(), importedBy: actorId })
        } else if (preview.kind === 'case-stage') {
          await requireAction('Cases', 'update-stage', actorId)
          const item = cases.find(record => record.caseNumber === row.values.caseNumber || record.id === row.values.caseNumber)!
          const product = catalogue.find(record => record.kind === 'products' && (record.id === item.productId || record.name === item.product && catalogue.some(bank => bank.kind === 'banks' && bank.id === record.parentId && bank.name === item.bank)))!
          const stage = stages.find(record => record.productId === product.id && record.name === row.values.stage)!
          if (item.coordinatorId !== actorId) throw new Error('Acting user is not the assigned Case Coordinator.')
          await caseRepository.update(item.id, { action: 'update-stage', actorId, value: stage.id })
        } else {
          const values = row.values
          await mockUsersRepository.createFromImport({ ...emptyUserDraft(), fullName: values.fullName, email: values.email, userType: values.userType, organization: values.organization, organizationScope: values.organizationScope as UserDraft['organizationScope'], officeBranch: values.officeBranch, department: values.department, team: values.team }, actorId)
        }
        success++
      } catch (cause) { results.push(`Line ${row.line}: ${cause instanceof Error ? cause.message : 'Import failed.'}`) }
    }
    if (preview.kind === 'attendance') writeArray(ATTENDANCE_KEY, attendance)
    const result: ImportResult = { id: crypto.randomUUID(), kind: preview.kind, at: new Date().toISOString(), actorId, success, failed: preview.rows.length - success, details: results }
    writeArray(IMPORT_RESULTS_KEY, [...readArray<ImportResult>(IMPORT_RESULTS_KEY), result]); return result
  },
}

File: backend/app/api/v1/csv_imports.py
"""Three controlled CSV intakes with server preview and atomic confirmation."""

import csv
from datetime import date, datetime, timedelta, timezone
from decimal import Decimal
from io import StringIO
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError

from app.api.v1.auth import Current, DB
from app.core.config import get_settings
from app.db.models import AttendanceImportRow, Case, CaseHistory, CsvImport, OrganizationUnit, PasswordSetupLink, ProductStage, User, UserType
from app.security.authorization import has_permission, require_permission
from app.security.passwords import new_token, token_hash
from app.api.v1.notifications import add_notification

router = APIRouter(prefix="/imports", tags=["CSV imports"])
Kind = Literal["attendance", "case-stage", "users"]
HEADERS = {
    "attendance": ["userEmail", "date", "status"],
    "case-stage": ["caseNumber", "stage"],
    "users": ["fullName", "email", "userType", "organization", "organizationScope", "officeBranch", "department", "team"],
}


class PreviewInput(BaseModel):
    kind: Kind
    csv: str = Field(min_length=1, max_length=2_000_000)


class ConfirmInput(BaseModel):
    preview_id: UUID


def parse_csv(kind: Kind, source: str) -> list[dict]:
    if "\0" in source or len(source.encode("utf-8")) > 2_000_000:
        raise HTTPException(status_code=422, detail="Choose a text CSV no larger than 2 MB")
    try:
        reader = csv.reader(StringIO(source.lstrip("\ufeff")), strict=True)
        header = next(reader, None)
        if header != HEADERS[kind]:
            raise HTTPException(status_code=422, detail=f"CSV headers must be: {', '.join(HEADERS[kind])}")
        records = list(reader)
    except csv.Error as exc:
        raise HTTPException(status_code=422, detail="CSV could not be parsed") from exc
    if len(records) > 1000:
        raise HTTPException(status_code=422, detail="Import at most 1000 rows at a time")
    return [{"line": index + 2,
             "values": dict(zip(HEADERS[kind], [cell.strip() for cell in cells])),
             "errors": ["Incorrect column count."] if len(cells) != len(HEADERS[kind]) else []}
            for index, cells in enumerate(records)]


def parse_id(value: str) -> UUID | None:
    try:
        return UUID(value)
    except ValueError:
        return None


async def resolve_org(values: dict[str, str], db: DB) -> dict[str, UUID | None]:
    scope = values["organizationScope"]
    if scope not in ("organization", "office", "department", "team"):
        raise ValueError("Invalid organization scope.")
    ids: dict[str, UUID | None] = {"organization_id": None, "office_id": None, "department_id": None, "team_id": None}
    prior: UUID | None = None
    fields = (("organization", "organization_id", "organizations"),
              ("officeBranch", "office_id", "offices"),
              ("department", "department_id", "departments"),
              ("team", "team_id", "teams"))
    for index, (label, key, kind) in enumerate(fields):
        name = values[label]
        if not name:
            if any(values[future_label] for future_label, _, _ in fields[index + 1:]):
                raise ValueError("Organization hierarchy is incomplete.")
            break
        unit = (await db.scalars(select(OrganizationUnit).where(
            OrganizationUnit.kind == kind,
            OrganizationUnit.parent_id == prior if prior else OrganizationUnit.parent_id.is_(None),
            func.lower(OrganizationUnit.name) == name.lower(),
            OrganizationUnit.active.is_(True),
        ))).one_or_none()
        if not unit:
            raise ValueError(f"Active {label} was not found in its parent scope.")
        ids[key] = unit.id
        prior = unit.id
    required = {"organization": "organization_id", "office": "office_id", "department": "department_id", "team": "team_id"}[scope]
    if ids[required] is None:
        raise ValueError("Organization scope needs a matching unit.")
    return ids


async def validate_rows(kind: Kind, rows: list[dict], db: DB, actor: User) -> tuple[list[dict], int, int]:
    seen: set[str] = set()
    matched = unmatched = 0
    for row in rows:
        values, errors = row["values"], row["errors"]
        if len(values) != len(HEADERS[kind]):
            continue
        if kind == "attendance":
            email = values["userEmail"].lower()
            user = (await db.scalars(select(User).where(func.lower(User.email) == email))).one_or_none()
            if not user:
                errors.append("User email was not found.")
            try:
                day = date.fromisoformat(values["date"])
                if day.isoformat() != values["date"]:
                    raise ValueError()
            except ValueError:
                day = None
                errors.append("Invalid date.")
            if values["status"] not in ("Present", "Absent", "Late"):
                errors.append("Invalid attendance status.")
            key = f"{email}:{values['date']}"
            existing = bool(user and day and (await db.scalars(select(AttendanceImportRow.id).where(
                AttendanceImportRow.user_id == user.id, AttendanceImportRow.date == day))).first())
            if key in seen or existing:
                errors.append("Duplicate attendance entry.")
            seen.add(key)
        elif kind == "case-stage":
            identifier = parse_id(values["caseNumber"])
            case = (await db.scalars(select(Case).where(
                (Case.case_number == values["caseNumber"]) | (Case.id == identifier) if identifier else Case.case_number == values["caseNumber"]
            ))).one_or_none()
            if not case:
                errors.append("Case was not found.")
                unmatched += 1
            else:
                matched += 1
                stage = (await db.scalars(select(ProductStage).where(
                    ProductStage.product_id == case.product_id,
                    ProductStage.name == values["stage"], ProductStage.active.is_(True)
                ))).one_or_none()
                if not stage:
                    errors.append("Stage is not active for this Case Product.")
                if case.status != "SM Approved" or not case.coordinator_id or not case.submitted_at or not case.locked_at:
                    errors.append("Case is not ready for stage tracking.")
                if case.coordinator_id != actor.id:
                    errors.append("Acting user is not the assigned Case Coordinator.")
                if stage and case.current_stage_id == stage.id:
                    errors.append("Case is already in this Stage.")
            if values["caseNumber"] in seen:
                errors.append("Duplicate Case row.")
            seen.add(values["caseNumber"])
        else:
            email = values["email"].lower()
            name = values["fullName"]
            if not name or len(name) > 120:
                errors.append("Full Name is required and must be 120 characters or fewer.")
            if len(email) > 320 or not email or "@" not in email or email.startswith("@") or email.endswith("@"):
                errors.append("Valid email is required.")
            if email in seen or (await db.scalars(select(User.id).where(func.lower(User.email) == email))).first():
                errors.append("Duplicate user email.")
            seen.add(email)
            user_type = (await db.scalars(select(UserType).where(
                func.lower(UserType.name) == values["userType"].lower(), UserType.active.is_(True)
            ))).one_or_none()
            if not user_type:
                errors.append("Active User Type was not found.")
            try:
                await resolve_org(values, db)
            except ValueError as exc:
                errors.append(str(exc))
    return rows, matched, unmatched


def preview_view(item: CsvImport) -> dict:
    rows = item.summary.get("rows", [])
    return {"id": str(item.id), "kind": item.kind, "rows": rows,
            "valid": sum(not row["errors"] for row in rows),
            "invalid": sum(bool(row["errors"]) for row in rows),
            "matchedCases": item.summary.get("matchedCases", 0),
            "unmatchedCases": item.summary.get("unmatchedCases", 0)}


def result_view(item: CsvImport, setup_links: list[str] | None = None) -> dict:
    details = [f"Line {row['line']}: {' '.join(row['errors'])}" for row in item.summary.get("rows", []) if row["errors"]]
    return {"id": str(item.id), "kind": item.kind, "at": (item.confirmed_at or item.created_at).isoformat(),
            "actorId": str(item.actor_id), "success": item.imported_count,
            "failed": item.error_count, "details": details + (setup_links or [])}


@router.post("/preview", status_code=201)
async def preview(payload: PreviewInput, current: Current, db: DB) -> dict:
    await require_permission(db, current.user, "Imports", payload.kind)
    rows = parse_csv(payload.kind, payload.csv)
    rows, matched, unmatched = await validate_rows(payload.kind, rows, db, current.user)
    item = CsvImport(kind=payload.kind, file_name="upload.csv", actor_id=current.user.id,
                     status="preview", row_count=len(rows), imported_count=0,
                     error_count=sum(bool(row["errors"]) for row in rows),
                     summary={"rows": rows, "matchedCases": matched, "unmatchedCases": unmatched})
    db.add(item)
    await db.commit()
    return preview_view(item)


@router.get("/history")
async def history(current: Current, db: DB) -> list[dict]:
    rows = (await db.scalars(select(CsvImport).where(CsvImport.status == "confirmed")
                             .order_by(CsvImport.confirmed_at.desc(), CsvImport.id.desc()).limit(100))).all()
    allowed = {kind for kind in HEADERS if await has_permission(db, current.user, "Imports", kind)}
    return [result_view(item) for item in rows if item.kind in allowed]


@router.post("/confirm")
async def confirm(payload: ConfirmInput, current: Current, db: DB) -> dict:
    item = (await db.scalars(select(CsvImport).where(CsvImport.id == payload.preview_id).with_for_update())).one_or_none()
    if not item or item.actor_id != current.user.id:
        raise HTTPException(status_code=404, detail="Import preview not found")
    if item.status != "preview":
        raise HTTPException(status_code=409, detail="Import was already confirmed")
    await require_permission(db, current.user, "Imports", item.kind)
    original = item.summary["rows"]
    if not any(not row["errors"] for row in original):
        raise HTTPException(status_code=422, detail="No valid rows are available to import")
    fresh = [{"line": row["line"], "values": row["values"], "errors": []} for row in original]
    fresh, _, _ = await validate_rows(item.kind, fresh, db, current.user)
    if any(not original[index]["errors"] and row["errors"] for index, row in enumerate(fresh)):
        raise HTTPException(status_code=409, detail="Source data changed; upload and preview the CSV again")
    if item.kind == "case-stage":
        await require_permission(db, current.user, "Cases", "update-stage")
    setup_links: list[str] = []
    now = datetime.now(timezone.utc)
    for row in original:
        if row["errors"]:
            continue
        values = row["values"]
        if item.kind == "attendance":
            user = (await db.scalars(select(User).where(func.lower(User.email) == values["userEmail"].lower()))).one()
            db.add(AttendanceImportRow(user_id=user.id, date=date.fromisoformat(values["date"]),
                                       status=values["status"], import_id=item.id))
        elif item.kind == "case-stage":
            identifier = parse_id(values["caseNumber"])
            case = (await db.scalars(select(Case).where(
                (Case.case_number == values["caseNumber"]) | (Case.id == identifier) if identifier else Case.case_number == values["caseNumber"]
            ).with_for_update())).one()
            stage = (await db.scalars(select(ProductStage).where(
                ProductStage.product_id == case.product_id,
                ProductStage.name == values["stage"], ProductStage.active.is_(True)
            ))).one()
            case.current_stage_id = stage.id
            case.stage_started_at = now
            case.stage_expected_duration_hours = stage.expected_duration_hours
            case.stage_due_at = now + timedelta(seconds=float(stage.expected_duration_hours * Decimal(3600)))
            case.updated_at = now
            db.add(CaseHistory(case_id=case.id, event=stage.name, actor_id=current.user.id,
                               detail=f"CSV import {item.id}; Expected Duration: {stage.expected_duration_hours} hours.",
                               stage_id=stage.id, stage_name=stage.name,
                               expected_duration_hours=stage.expected_duration_hours, at=now))
            await add_notification(db, case.case_owner_id, "case-stage",
                                   f"{case.case_number}: {stage.name}", current.user.id, case.id)
        else:
            user_type = (await db.scalars(select(UserType).where(
                func.lower(UserType.name) == values["userType"].lower(), UserType.active.is_(True)
            ))).one()
            ids = await resolve_org(values, db)
            user = User(email=values["email"].lower(), full_name=values["fullName"],
                        user_type_id=user_type.id, organization_scope=values["organizationScope"],
                        active=False, **ids)
            db.add(user)
            await db.flush()
            token = new_token()
            db.add(PasswordSetupLink(user_id=user.id, token_hash=token_hash(token),
                                     created_by_id=current.user.id, created_at=now,
                                     expires_at=now + timedelta(minutes=get_settings().setup_link_minutes)))
            setup_links.append(f"Line {row['line']}: /set-password?token={token}")
    item.status = "confirmed"
    item.imported_count = sum(not row["errors"] for row in original)
    item.error_count = sum(bool(row["errors"]) for row in original)
    item.confirmed_at = now
    try:
        await db.commit()
    except IntegrityError as exc:
        await db.rollback()
        raise HTTPException(status_code=409, detail="Import conflicted with current data; preview again") from exc
    return result_view(item, setup_links)
